### PR TITLE
XCOMMONS-2568: Improve comment handling in HTMLCleaner

### PR DESCRIPTION
**Pull request by michitux**
Jira issue: https://jira.xwiki.org/browse/XCOMMONS-2568

### CODE REVIEW

**Review by tmortagne — APPROVED** [review]

File: xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/internal/html/DefaultHTMLCleaner.java [hunk 1]
```diff
@@ -267,6 +267,9 @@ private CleanerProperties getDefaultCleanerProperties(HTMLCleanerConfiguration c
 
         defaultProperties.setDeserializeEntities(true);
 
+        // Omit comments in restricted mode to avoid any potential parser confusion.
```
[the remark this answers is not in the document]
Comment (michitux) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/internal/html/DefaultHTMLCleaner.java:270]: At least from my side, there is no intention to remove this as any part of the filtered HTML that contains unfiltered input is potentially dangerous/a candidate for parser confusion. Comments, style and script elements seem to be frequently found ingredients in successful attacks against good sanitizers. We're already removing style and script elements, so removing comments seems like a good defense against future attacks. If we didn't want to remove comments, we don't need this change. For further reference, see also the discussion in the [forum](https://forum.xwiki.org/t/removing-html-comments-in-restricted-html-filtering/11302).

[the remark this answers is not in the document]
Comment (vmassol) [reply xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/internal/html/DefaultHTMLCleaner.java:270]: Ok, sorry for not following closely. Then I believe we need a more elaborate comment to explain why, i.e. copy your answer as a comment maybe ;)

File: xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java [hunk 2]
```diff
@@ -327,6 +327,40 @@ void restrictedAttributesAndTags() throws Exception
...
+        // Actually, just the comment should be removed but due to erroneous parsing in HTMLCleaner, the whole thing
+        // is treated as a comment.
+        assertHTML("", "<!--> <a href=\"#\">no comment</a>");
+        assertHTML("", "<!---> <a href=\"#\">no comment</a>");
+    }
+
+    @Test
+    void comments()
+    {
+        assertHTML("<!--My favorite operators are > and <!-->", "<!--My favorite operators are > and <!-->");
```
Comment (manuelleduc) [inline xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java:352]: On small improvement I can think of is to use parameterized tests. The main interest it that the test will not stop when the first assertion fail, but will report individual failures for each assertions.

Comment (michitux) [reply xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java]: I've changed the two new tests into parameterized tests.

**Pull request merged**